modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +52,135 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def parseMultipleDataTransactions: F[DataTransactions] = {
+        implicit val dataTransactionRequest: EntityDecoder[F, DataTransactions] = dataApplication.requestDecoder
+        req.as[DataTransactions]
+      }
+
+      def parseSingleDataTransaction: F[DataTransactions] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+          .map(NonEmptyList.of(_))
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions, isSingleUpdate: Boolean): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+            }
+          }.flatMap { results =>
+            val errors = results.collect { case Left(_) => InvalidSignature.toApplicationError }
+            val validTransactions = results.collect { case Right(hashed) => hashed }
+
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
+              )
+            } else if (validTransactions.isEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("No valid data transactions found"))
+              )
+            } else {
+              NonEmptyList.fromList(validTransactions.map(_.hash)) match {
+                case Some(hashes) =>
+                  dataTransactionsQueue
+                    .offer(NonEmptyList.fromListUnsafe(validTransactions.map(_.signed)))
+                    .flatMap { _ =>
+                      if (isSingleUpdate) {
+                        Ok(("hash" ->> hashes.head.value) :: HNil)
+                      } else {
+                        Ok("hashes" ->> hashes.map(_.value) :: HNil)
+                      }
+                    }
+                case None =>
+                  BadRequest(
+                    Json.obj("error" -> Json.fromString("No valid data transactions found"))
+                  )
               }
+            }
           }
         }
-        .handleUnknownError
+
+      parseMultipleDataTransactions
+        .flatMap(handleDataUpdates(_, isSingleUpdate = false))
```

Comment:
instead of `isSingleUpdate` parameter, just create two definitions with the same name (overload the signature) - one with single element, second one with a list of elements

Reply:
👍🏼